controller/controller.go

```diff
@@ -698,6 +704,96 @@ func (s *Service) heartbeatController(ctx context.Context, addr *url.URL) {
 	}
 }
 
+func (s *Service) watchModuleChanges(ctx context.Context, sendChange func(response *ftlv1.PullSchemaResponse) error) error {
+	moduleSchemas := map[string]schema.Module{}
+	deployments, err := s.dal.GetActiveDeployments(ctx)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	for i, deployment := range deployments {
+		moduleSchemas[deployment.Schema.Name] = *deployment.Schema
+		//nolint:forcetypeassert
+		err := sendChange(&ftlv1.PullSchemaResponse{
+			DeploymentKey: deployment.Key.String(),
+			Schema:        deployment.Schema.ToProto().(*pschema.Module),
+			More:          i < len(deployments)-1,
+			ChangeType:    ftlv1.DeploymentChangeType_DEPLOYMENT_ADDED,
+		})
+		if err != nil {
+			return errors.WithStack(err)
+		}
+	}
+
+	for {
+		dbDeployments := map[string]dal.Deployment{}
+		var changesToSend []*ftlv1.PullSchemaResponse
+
+		deployments, err := s.dal.GetActiveDeployments(ctx)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		for _, deployment := range deployments {
+			dbDeployments[deployment.Schema.Name] = deployment
+			if current, ok := moduleSchemas[deployment.Schema.Name]; ok {
+				if !proto.Equal(current.ToProto(), deployment.Schema.ToProto()) {
+					//nolint:forcetypeassert
+					changesToSend = append(changesToSend, &ftlv1.PullSchemaResponse{
+						DeploymentKey: deployment.Key.String(),
+						Schema:        deployment.Schema.ToProto().(*pschema.Module),
+						ChangeType:    ftlv1.DeploymentChangeType_DEPLOYMENT_CHANGED,
+					})
+				}
+			} else {
+				//nolint:forcetypeassert
+				changesToSend = append(changesToSend, &ftlv1.PullSchemaResponse{
+					DeploymentKey: deployment.Key.String(),
+					Schema:        deployment.Schema.ToProto().(*pschema.Module),
+					ChangeType:    ftlv1.DeploymentChangeType_DEPLOYMENT_ADDED,
+				})
+			}
+			moduleSchemas[deployment.Schema.Name] = *deployment.Schema
+		}
+
+		var keysToRemove []string
+		for name, moduleSchema := range moduleSchemas {
+			if dbDeployment, ok := dbDeployments[name]; !ok {
+				keysToRemove = append(keysToRemove, name)
```

Comment:
You can delete keys while iterating over the map - Go will handle this correctly.